### ISSUE TITLE
False positive for global-variable-not-assigned when using with

### ISSUE THREAD

**Issue by andy-maier** [PyCQA/pylint#5073]
### Bug description

```python
# File a.py
"""
The "global-variable-not-assigned" check does not take into account that a "with"
statement can very well change the global variable.
For example, when using a global condition variable by multiple threads:
"""

import threading

COND = threading.Condition()


def func1():
    "Function that is called in one thread"
    # pylint: disable=global-statement
    global COND
    with COND:
        COND.notify()
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:16:4: W0602: Using global for 'COND' but no assignment is done (global-variable-not-assigned)
```

### Expected behavior

Pylint realizes that "with" may change the global variable and thus counts it like it was an assignment, and thus does not raise the issue in this case.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.0
Python 3.9.7 (default, Sep  3 2021, 12:45:31) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

macOS

### Additional dependencies

Output of pipdeptree:
pylint==1.6.4
  - astroid [required: >=1.4.5,<1.5.0, installed: 1.4.9]
    - lazy-object-proxy [required: Any, installed: 1.4.2]
    - six [required: Any, installed: 1.15.0]
    - wrapt [required: Any, installed: 1.11.2]
  - backports.functools-lru-cache [required: Any, installed: 1.6.4]
  - configparser [required: Any, installed: 4.0.2]
  - isort [required: >=4.2.5, installed: 4.2.15]
  - mccabe [required: Any, installed: 0.6.1]
  - six [required: Any, installed: 1.15.0]

**Comment by timmartin** [comment]
Is it true that the `with` statement is assigning to the variable? It should only read and execute the `__enter__` and `__exit__` methods on the object, and neither of these should be able to bind the global variable to a new object. Of course it's possible for `with` statement to mutate the contents of the object, but that's true of any method call on the object, and doesn't necessitate `global`.

It's worth noting that variables qualified by `global` are explicitly disallowed from being used as a _target_ of a `with` statement, presumably because in that case the variable can be reassigned to a new object, and it would create unnecessary complexity to handle this.

**Comment by nickdrozd** [comment]
Is the global statement required here? What happens if it's removed?

**Comment by andy-maier** [comment]
@nickdrozd @timmartin 
Sorry for the delay.

The `global` statement is in fact not needed, because the global name is not bound again in the function. In our real code, that is also the case. Thanks for the hint, so I can simplify our code there.

However, I still think raising "global-variable-not-assigned error" is not right here. It is assigned (in the global scope) and it cannot be reassigned by the `with` statement.

**Comment by timmartin** [comment]
I don't understand. You're saying that the `global` is not needed, but also claiming that the Pylint warning telling you that `global` is not needed is a false positive? Surely if the `global` is unnecessary then Pylint is doing exactly the right thing by telling you this?

Perhaps the name `global-variable-not-assigned` is causing some confusion. I think the point of the warning is to say that it isn't assigned, therefore global is unnecessary, but I can see how that string could be read as "variable not assigned", which I don't think is intended.

**Comment by mbyrnepr2** [comment]
Right, I think a documentation update could be made to clarify it. Currently the message is not explicit about where the assignment is expected (in the same block as the global statement).

**Comment by timmartin** [comment]
Perhaps change the message to something like `Unnecessary use of global for 'COND' not assigned in scope`? It's a bit clunky, but we want to fit the message on a single line. Or possibly even `Unnecessary use of global for 'COND'`? It sounds like the reference to assignment is making things less clear, not more clear.

**Comment by mbyrnepr2** [comment]
I agree and I think your second example is a good clear and simple way to go.
